File: 2021/csharp/Day1.cs
namespace AdventOfCode.CSharp.Y2021;

using Microsoft.FSharp.Collections;

class Day1 : AdventOfCode.CSharp.RobotElf
{
    public Day1() : base(1) {}

    public override object Part1()
    {
        var numbers = Input.Select(int.Parse);

        return Increasing(numbers);
    }

    public override object Part2() 
    {
        var windows = 
            SeqModule
                .Windowed(3, Input.Select(int.Parse))
                .Select(s => s.Sum());

        return Increasing(windows);
    }

    int Increasing(IEnumerable<int> numbers)
    {
        var last = int.MaxValue;
        var total = 0;
        foreach (var n in numbers)
        {
            if (n > last) total += 1;
            last = n;
        }
        return total;
    }  
}

File: 2021/fsharp/day1.fs
namespace AdventOfCode.FSharp.Y2021

open System

module Day1 =

    let parse input =
        input
        |> Seq.map Int32.Parse
        |> Seq.toArray


    let part1 input =
        let x = 
            input
            |> parse

        Array.allPairs x x 
        |> Seq.filter (fun (a, b) -> a <> b && (a + b) = 2020)
        |> Seq.map (fun (a, b) -> a * b)
        |> Seq.head
        |> bigint

    let part2 input = 
        let x = 
            input
            |> parse

        Array.allPairs x x
        |> Array.allPairs x
        |> Seq.filter (fun (a, (b, c)) -> a <> b && a <> c && (a + b + c) = 2020)
        |> Seq.map (fun (a, (b, c)) -> a * b * c)
        |> Seq.head
        |> bigint